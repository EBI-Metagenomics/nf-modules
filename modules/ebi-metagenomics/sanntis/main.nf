process SANNTIS {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'https://depot.galaxyproject.org/singularity/sanntis:0.9.4.1--pyhdfd78af_0'
        : 'biocontainers/sanntis:0.9.4.1--pyhdfd78af_0'}"

    input:
    tuple val(meta), path(interproscan), path(gbk), path(faa)

    output:
    tuple val(meta), path("*_sanntis.gff.gz"), emit: gff
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    if (faa && gbk) {
        error("SanntiS supports either a GBK or a FAA as the secondary input.")
    }
    if (!gbk && !faa) {
        error("SanntiS needs either a GBK or FAA as input.")
    }

    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def is_ips_compressed = interproscan.name.endsWith(".gz")
    def  interproscan_file = interproscan ? interproscan.name.replace(".gz", "") : is_ips_compressed

    // Handle the GBK or FAA
    def input_arg = ""
    def input_file = ""
    def decompress = ""
    def clean_decompressed = ""
    if (gbk) {
        input_file = gbk.name
        if (gbk.name.endsWith(".gz")) {
            input_file = input_file.replace(".gz", "")
            decompress = "gzip -c -d ${gbk} | grep -v \"/protein_id=\" > ${input_file}"
            clean_decompressed = "rm -f ${input_file}"
        }
        input_arg = "${input_file}"
    }
    if (faa) {
        input_file = faa.name
        if (faa.name.endsWith(".gz")) {
            input_file = input_file.replace(".gz", "")
            decompress = "gzip -c -d ${faa} > ${input_file}"
            clean_decompressed = "rm -f ${input_file}"
        }
        input_arg = "--is_protein ${input_file}"
    }
    """
    if [ "${is_ips_compressed}" == "true" ]; then
         gzip -c -d ${interproscan} > ${interproscan_file}
    fi

    ${decompress}

    sanntis \\
        --ip-file ${interproscan_file} \\
        --outfile ${prefix}_sanntis.gff \\
        --cpu ${task.cpus} \\
        ${args} \\
        ${input_arg}

    gzip ${prefix}_sanntis.gff

    # Purge the decompressed files to save storage
    ${clean_decompressed}
    if [ "${is_ips_compressed}" == "true" ]; then
         rm -f ${interproscan_file}
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        SanntiS: \$(sanntis --version | sed "s/SanntiS\\ //g")
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_sanntis.gff
    gzip ${prefix}_sanntis.gff

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        SanntiS: \$(sanntis --version | sed "s/SanntiS\\ //g")
    END_VERSIONS
    """
}
