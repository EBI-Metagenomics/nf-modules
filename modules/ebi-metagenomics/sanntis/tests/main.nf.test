nextflow_process {

    name "Test Process SANNTIS"
    script "../main.nf"
    process "SANNTIS"

    tag "modules"
    tag "modules_ebimetagenomics"
    tag "sanntis"

    test("interproscan tsv and faa") {

        when {
            process {
                """
                input[0] = [
                    [ id:'BGC0001472' ], // meta map
                    file("https://raw.githubusercontent.com/Finn-Lab/SanntiS/refs/heads/master/test/files/BGC0001472.fna.prodigal.faa.gff3", checkIfExists: true),
                    [],
                    file("https://raw.githubusercontent.com/Finn-Lab/SanntiS/refs/heads/master/test/files/BGC0001472.fna.prodigal.faa", checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("interproscan tsv and faa both compress") {

        when {
            process {
                """
                input[0] = [
                    [ id:'BGC0001472' ], // meta map
                    file("${moduleDir}/tests/fixtures/BGC0001472.fna.prodigal.faa.gff3.gz", checkIfExists: true),
                    [],
                    file("${moduleDir}/tests/fixtures/BGC0001472.fna.prodigal.faa.gz", checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("interproscan tsv and faa - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'BGC0001472' ], // meta map
                    file("https://raw.githubusercontent.com/Finn-Lab/SanntiS/refs/heads/master/test/files/BGC0001472.fna.prodigal.faa.gff3", checkIfExists: true),
                    [],
                    file("https://raw.githubusercontent.com/Finn-Lab/SanntiS/refs/heads/master/test/files/BGC0001472.fna.prodigal.faa", checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
