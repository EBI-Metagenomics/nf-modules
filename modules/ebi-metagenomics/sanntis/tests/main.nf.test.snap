{
    "interproscan tsv and faa both compress": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "BGC0001472"
                        },
                        "BGC0001472_sanntis.gff.gz:md5,ea6f1c2d950ca684a5e8ca77d2bf773e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,5fc42785fbe619f4827b419532cde346"
                ],
                "gff": [
                    [
                        {
                            "id": "BGC0001472"
                        },
                        "BGC0001472_sanntis.gff.gz:md5,ea6f1c2d950ca684a5e8ca77d2bf773e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5fc42785fbe619f4827b419532cde346"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T10:07:37.995879"
    },
    "interproscan tsv and faa": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "BGC0001472"
                        },
                        "BGC0001472_sanntis.gff.gz:md5,ea6f1c2d950ca684a5e8ca77d2bf773e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,5fc42785fbe619f4827b419532cde346"
                ],
                "gff": [
                    [
                        {
                            "id": "BGC0001472"
                        },
                        "BGC0001472_sanntis.gff.gz:md5,ea6f1c2d950ca684a5e8ca77d2bf773e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5fc42785fbe619f4827b419532cde346"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-03T22:44:54.79901"
    },
    "interproscan tsv and faa - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "BGC0001472"
                        },
                        "BGC0001472_sanntis.gff.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,5fc42785fbe619f4827b419532cde346"
                ],
                "gff": [
                    [
                        {
                            "id": "BGC0001472"
                        },
                        "BGC0001472_sanntis.gff.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5fc42785fbe619f4827b419532cde346"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-03T22:45:03.056725"
    }
}